test/msw-api/setup-worker/use.test.ts

```diff
@@ -108,3 +109,45 @@ test('returns a mocked response from a one-time request handler override only up
 
   await runtime.cleanup()
 })
+
+test('returns a mocked response from a one-time request handler override only upon first request match with parallel requests', async () => {
+  const runtime = await runBrowserWith(path.resolve(__dirname, 'use.mocks.ts'))
+
+  await runtime.page.evaluate(() => {
+    const { msw } = window
+
+    msw.worker.use(
+      msw.rest.get('/book/:bookId', function oneTimeOverride(req, res, ctx) {
+        return res.once(ctx.json({ title: 'One-time override' }))
+      }),
+    )
+  })
+
+  const bookPromise = runtime.request({
+    url: runtime.makeUrl('/book/abc-123?id=book'),
+    responsePredicate(res, url) {
+      return match(url, res.url()).matches
+    },
+  })
+
+  const anotherBookPromise = runtime.request({
+    url: runtime.makeUrl('/book/abc-123?id=anotherBook'),
+    responsePredicate(res, url) {
+      return match(url, res.url()).matches
+    },
+  })
+
+  const bookResponse = await bookPromise
+  const bookStatus = bookResponse.status()
+  const bookBody = await bookResponse.json()
+  expect(bookStatus).toBe(200)
+  expect(bookBody).toEqual({ title: 'One-time override' })
+
+  const anotherBookResponse = await anotherBookPromise
+  const anotherBookStatus = anotherBookResponse.status()
+  const anotherBookBody = await anotherBookResponse.json()
+  expect(anotherBookStatus).toBe(200)
+  expect(anotherBookBody).toEqual({ title: 'Original title' })
```

Comment:
One thing that was tricky to create the test for the `setupWorker` case was that the test would get the wrong title here (`One-time override`), instead of the expected one (`Original title`). 

The reason for that was because `anotherBookPromise` was being set to the same promise as `bookPromise` since they had the same URL and the default `responsePredicate` inside `runtime.request()`.

So to overcome this, I added a query param to the URL with a different value for each request.

Then I passed a custom `responsePredicate` that matches the full URL, including the query params, which is different than the default one, which removes any query params before checking for matches.

I'm not sure if there is a better way to do that. Suggestions?

Reply:
Considering that I was wondering if would be clearer if the URL was set to something like that:

```ts
runtime.makeUrl('/book/abc-123?promiseId=anotherBook')
```

```ts
runtime.makeUrl('/book/abc-123?requestId=anotherBook')
```

It would not make a difference for the test, I'm just thinking if it would make the test more readable.

Reply:
That's a great concern about the internal response awaiting utility we use for testing. There can be cases where multiple requests are made to the same URL, and they should resolve properly. 

Adding a query parameter looks like a workaround rather than a solution. Since the issue lies in how we map request/response in tests, I suggest we fix it there. While adding distinct information like different query parameter will work, I think it creates a confusing test because requests now contain data they don't need to showcase a usage example. That data is used solely to compensate for the internal testing setup.

Reply:
I will provide the fix for this issue so you don't have to worry about the right response being awaited.

Reply:
This is addressed in https://github.com/mswjs/msw/pull/584.